### ISSUE TITLE
Making the SF102 update using the BLE factory bootloader (no need to open the case)

### ISSUE THREAD

**Issue by geeksville** [OpenSource-EBike-firmware/SW102_LCD_Bluetooth#35]
I think ya'll will love this thread (@casainho and @lowPerformer especially).  So eyebyesickle asked me for a smaller build because he tried to use our appload with the factory bootloader and it was complaining.  I made the smaller build and it still complained.  Some super interesting things followed (and huge thanks to eyebyesickle for making it possible):

* He mentioned that the mfg said their stock bootloader isn't checking signatures on apploads
* They used version 9 of the SDK.

So I just went trolling through the docs (and I bet @lowPerformer knows this better) but - sure enough version 9 of the nordic bootloader protocol doesn't check app signatures against a private key.  This is **super** great news.

The docs on their rev 9 SDK are much more rough ( https://developer.nordicsemi.com/nRF5_SDK/nRF51_SDK_v9.x.x/ ).  But they include the source for their bootloader.

I think the only thing we need to do to transition these factory devices is:

* Make a second OTA package called sw102-xxx-firsttime.zip.  But we need to use the 0.5.2 version of the dfutool app.  The version we've been using from our newer SDKs will not work - file formats changed.
* When running that tool we need to tell it to include the bootloader and the softdevice (so we can transition this old firmware to a soft device with the features we need).  Of course we will need to tell users to DEFINITELY not power down during this bootloader update. (the memory map changed for later soft devices, and the bootloader protocol changed as well)
* We might also be able to include the appload in firttime.zip (I haven't checked), but if not the user can then just rerun the DFU app on their phone to install our normal OTA file.

Do any of ya'll have a SF102 that still has the factory software on to try this with?  eyebyesickle is happy to try, but I'd prefer to try it when we can run with bluetooth logging turned on in the DFU android app, to watch the process and see/fix whatever errors occur.

**Comment by geeksville** [comment]
ooh - eyebyesickle just told me he's loaning me a virgin SF102 to try this with.  So when it gets here (2-3 days probably) I'll try this out.

**Comment by geeksville** [comment]
Also, this combined with setting the option byte on the TSDZ2 to turn on their built in serial bootloader would probably allow a total switch to this open-source software without even an ST-LINK.  

Assuming someone (eyebyesickle?) either gets TSDZ2s with that option byte set, or they briefly attach a ST-LINK once to set that byte (and optionally load the new appload).  Future updates of the TSDZ2 could be controlled by SF102.

**Comment by geeksville** [comment]
Update: We want to use 0.5.3 of nrfutil - not 0.5.2.  They fixed something about the python in 0.5.3 but file formats still compatible with 0.5.2.  I'm going to try building a quick experiment

https://github.com/NordicSemiconductor/pc-nrfutil/commits/0_5_3

**Comment by eyebyesickle** [comment]
Wow you are taking this to the next level. I'll get the SW102 out to you today. Let me know if I can do anything else to help... I lack expertise but try to make up for it with determination so maybe I can get it done haha.

I think I can get the option byte thing worked out, but of course, just as with the pre-enabled bluetooth SW102.... I would have to bulk order some US stock, and arrange Chinese warehouse stock to make this price reasonable for everyone... When push comes to shove I guess the first step would be getting the option byte we want to use, and then negotiating the implementation at the production line or otherwise. 

Will really be great once we get this all settled... The OSF will truly be plug and play user friendly! I would guess under 50% of people who want the OSF would dare to try it now, given the tasks that must be undertaken to install...

**Comment by lowPerformer** [comment]
Indeed interesting. I wasn't able to start a bootloader on a stock SW102 no matter which buttons I pressed for how long ;)
Unfortunately I don't have a original SW102 at hand and can't test...

**Comment by eyebyesickle** [comment]
@lowPerformer - You wouldn't be able to. Only me (Eco Cycles) and EggRider get the SW102 with bluetooth pre-enabled. I also have a batch coming in to the US within a couple weeks and am sending Mark(Rydon on ES) @ Electrify Bike some too... we like to share here, as this is for everyone. With the ones I get, you don't even have to press buttons, you just fire up the NRFtool and initiate the firmwareupdate, and the bluetooth logo pops up... 

Once we get everything sorted, I will get a bunch of the SW102 w/ bluetooth enabled at a Chinese warehouse so everyone will have access to reasonably priced stock... sounds like I might hold off just for a bit, until we verify the hardware changes we want

**Comment by geeksville** [comment]
screw caution, I just sent an untested firsttime.zip package to @eyebyesickle.  It might work ;-)

**Comment by geeksville** [comment]
> Only me (Eco Cycles) and EggRider get the SW102 with bluetooth pre-enabled.

Oh I see!  That makes sense and is totally fair.

re: the tsdz2 option byte preprogramming
I don't even think it is really that critical - with a little jig based on the casainho instructions plus an STLINK you can program the opensource software and set the option byte to allow future software update over serial (i.e. from the SF102).  So for each motor, just briefly hook it up to the spedo connector - program and put back in box.

**Comment by geeksville** [comment]
btw @lowPerformer your bootloader didn't do anything to turn off the 'feature' of letting the DFU tool ask to have the bootloader upgraded did it? (I assume no)  

I agree that feature is a little scary (but useful in rare emergencies) but if this test works and @eyebyesickle ever wants to return his test unit back to the factory bootloader we'd want that.

**Comment by lowPerformer** [comment]
"our" bootloader can update everything, including itself.
From the SDK docu:
"Note that, depending on the combination of images, the created zip file might contain two firmware packages (one for the bootloader and/or SoftDevice and one for the application) and the DFU process will be carried out in two steps."

But I really don't know if downgrading will work. They changed a lot in bootloader structure from v9 to v12.3 SDK.

**Comment by lowPerformer** [comment]
For first tests I would not update the bootloader.

**Comment by geeksville** [comment]
I would also desire that.  But two problems with that approach (for
actually running an appload):

* The existing soft device is ancient.
* Their bootloader is a different size (probably larger given they draw
graphics) than ours.

But good point.  I'll start with just linking our app so that it expects to
run from the location theirs starts. And at least get it to load a version
of our app.  Then switch to getting it to install our bootloader (with a
new soft device)

(Sent from a phone - please ignore typos)

**Comment by geeksville** [comment]
@lowPerformer btw - do you know if softdevices are supposed to be API compatible between SDK versions? i.e. could we just leave the soft device and bootloader they used and (if appload adjusted for the appropriate start address) and just update the appload?

**Comment by lowPerformer** [comment]
Uff, I doubt but not sure.

* Larger bootloader is not a problem. It sits at the end of flash while App+Softdevice starts from beginning. Sure app should not be too large.

So I would try to update a package with App and Softdevice and leave bootloader for first try.

**Comment by casainho** [comment]
My notes:
I bought a SW102 from PSWPower and the bootloader is activated while firmware runs, there was DFU but I hadn't any firmware zip package to test at that time. Yes, the bootloader can runs simultaneously with firmware.

On more recent SDKs at least, the bootloader, softdevice and app can be updated by the DFU itself and should be a secure process assuming encyrpttion keys are enabled and so only a trusted source for that pieces of firmware do exist. See that if they did disable key check, that means anyone will be able to brick the device be flashing a wrong bootloader.

**Comment by casainho** [comment]
And maybe their bootloader starts at boot if it doesn't find a valid app image or if the update firmware process breakst at the middle.

The bootloader can take a lot of space, can be configured to dual bank firmware update and so firmware can be only half of the size... I mean, small.

**Comment by geeksville** [comment]
@casainho their bootloader/appload was built with version 9 of the SDK and that version didn't check keys for updates.  I double checked against an appload upgrade file @eyebyesickle sent me (for the stock closed source app) and yep, the .dat file for that file is too small to also contain a key.

**Comment by geeksville** [comment]
> The bootloader can take a lot of space, can be configured to dual bank firmware update and so firmware can be only half of the size... I mean, small.

Yep - I bet that's why the new experimental appload I sent @eyebyesickle didn't work (even though I built it for a sdk 9 bootloader.  Also he mentioned their bootloader draws graphics (so more code).

When the test unit he sent me I'll first try not messing with the bootloader, but then if we have to install our new better bootloader I bet that will work.

**Comment by geeksville** [comment]
Heh.  I can already tell debugging the DFU flow using the Android nordic semi bluetooth logging will be an immense pain in the ass.  So I just ordered this little $10 devboard, apparently it works with nrfutil as a linux to BLE endpoint - and provides nice visibility of what is happening when a host asks these devices to update: https://www.mouser.com/ProductDetail/Nordic-Semiconductor/nRF52840-Dongle

**Comment by casainho** [comment]
I did power on the SW102 I have here that I bought from PSWPower and has the original firmware. Bluetooth does work as seen on the NRFconnect app. Please see what you can find... can we find the company looking at the IDs on the screenshots??

The device name is "Tophmi_8904". Topology Human Machine Interface, since the company is Topology Tech?? -- I don't see it listed on Bluetooth site..... or maybe I didn't search correctly??

See the screenshots I took, here: https://www.endless-sphere.com/forums/viewtopic.php?f=30&t=99698&start=75#p1464660

https://postimg.cc/F7cd35HR

https://postimg.cc/1gyqyCPz

https://postimg.cc/0rhKz2xF

**Comment by geeksville** [comment]
Okay - I poked around with the unit I received from @eyebyesickle (thanks!).  I also see Tophmi_8904 when the unit is in its appload.  I think I understand a little more now:

The unit used SDK 9.  But I've tried all combinations of buttons while booting and nothing will force the device to stay in the bootloader for DFU.  I bet what happened is they read these old [instructions](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v11.0.0%2Fbledfu_appextending.html).  and didn't bother to implement the (IMO) super important feature of forcing the bootloader to stay around if a button is pressed.  Instead they only implemented "The application on the DFU target supports entering DFU mode. In this case, the DFU controller can trigger the application to reset the device and enter DFU mode, which is referred to as buttonless update."

This was a really dumb/dangerous decision (though a nice feature **if you also implement the button on boot feature**) it depends on the appload to be healthy enough to 'help' with doing a bootload upgrade.  Which is pretty scary for reverse engineering because if we shoot down a bad appload the device would be dead (at least until cutting it open).

* Since I don't want to risk asking the device to install our bootloader yet, I think we should add a DFU endpoint to our appload.  This would allow us to try updating just the appload on these devices without risking borking the bootloader or soft device (and still have a way to keep updating the device in the future - as opposed to just once).  Instructions [here](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v11.0.0%2Fbledfu_appextending.html)).  I'm currently busy with the 850C/SW102 merge for the next 2-3 days (to unblock @casainho so we can start collaborating again on that branch).  After I have this done I can do this or perhaps @lowPerformer (I think the most softdevice/bootloader aware person here?) has time and interest to do this before then?

I also looked at the .dat files @eyebyesickle sent me for their updates and it indicates they depend on SD version S110 8.0.0.

**Comment by lowPerformer** [comment]
I think backporting to SDK 10 is not the way to go.
Better to try
1) updating App + SD130 with installed Bootliader. This should be possible out of the box imho.
2) Update everything

Even if 2 should not be possible, I can life with the older bootloadet version...

**Comment by casainho** [comment]
Wow!! Very good work, I am learning :-)

**Comment by lowPerformer** [comment]
Uff, thats a story to tell your grandchildren besides a crackling campfire at Christmas Eve ;)

I hope this is worth it. Please call me a low performer but I didn't get the point about that "special" SW102 completely yet.
1) Who sells this and when is it available
2) Availability (international)?
2 1/2) How much?
3) eyebyesickle told the manufacturer can put our firmware on the devices? So we just have to send them the already available full_hex.

**Comment by casainho** [comment]
I should receive one SW102 from PSWPower in 4 weeks, then I will be able to test.

I wish we could block easy access to update firmware, bootloader and SD. I see as a potential security as anyone will be able to render useless the SW102.

**Comment by cookejames** [comment]
I would be keen to volunteer as I've been waiting for this to be working to switch from the marcoq firmware and I am very keen on not having to cut the unit open if it is avoidable. Is the version from PSWPower good to buy?

**Comment by geeksville** [comment]
Interesting!  Your device might be fine, the bootloader doesn't use the
screen.  If you hold down m and power for eight seconds then go look at nrf
toolbox dfu.  Do you see "sw102 dfu" listed as a device?

(Sent from a phone - please ignore typos)

**Comment by Pscope** [comment]
unfortunately the same happened to me as well. 

I chose DFU and sent the sw102-transitional-0.19.13.zip file, all transferred as expected, I restarted then unable to boot into DFU again tried all manner of combinations.

**Comment by geeksville** [comment]
Ok alas.   I'll ponder a bit and reply tonight 😃

(Sent from a phone - please ignore typos)

**Comment by Pscope** [comment]
Just a thought. But when i clicked on DFU option in NRF connect the top option was a update zip file file which i chose as thats how it was packaged but there is an option to just upload a bootloader image. would this maybe be an option?

**Comment by casainho** [comment]
I am pretty sure @geeksville saw that option. That would be no different from the result you had, it would not work anyway.

**Comment by geeksville** [comment]
Yes alas.  There must be something still missing.  I'll ponder more
tonight.  It is tricky when you only get one chance to try. 😄

(Sent from a phone - please ignore typos)

**Comment by Pscope** [comment]
I have another unit arriving in the next day or two, so happy to try again if you find anything you think may help and want to try again.

**Comment by geeksville** [comment]
@Pscope ooh yes.  I do have another thing to try (if you'll be having to open it anyways if it fails).  I'm checking now on my board and will post a upgrade package for you in a few hrs.

**Comment by geeksville** [comment]
okay - I have low confidence this one is much better than the last, but if you are willing to try it I'd love to hear a report back.  Godspeed.

Please, only @Pscope  try this.  [sw102-transitional-0.19.14.zip](https://github.com/OpenSource-EBike-firmware/SW102_LCD_Bluetooth/files/3560621/sw102-transitional-0.19.14.zip)

**Comment by Pscope** [comment]
Okay well that did not work either, however. I may have made a fundamental mistake. Having rered your instructions again. I realised i was not using the same nrf app as you. I was using NRF/Connect and not NRF/toolbox. The process looked identical apart from when you select the file there is no option to choose 'All' it just sends the file.

Maybe that was wrong? 

Unfortunately i am all out of stock screens now, and no more on order at the moment, but whoever tries this next i guess should use the android app nrf/toolbox. Sorry maybe a rookie mistake but it appeared to be exactly the same apart from that small step.

**Comment by geeksville** [comment]
Good progress!  (thanks to @eyebyesickle sending me another unit).  Two updates included here so I don't misplace them.  I'll resume tomorrow night with the next tasks.

# Update 1:
yay! I think this bluetooth only initial thing will finally work. Eyebysickle sent me one more SW102 (really nice of him) so I could try a new approach. I when I tried before I could only really "test" my approach once because the factory SW102s have the read-protection bit set in their config flash - so you aren't supposed to be able to read anything from jtag and all writes must first erase the entire flash. (cc @lowPerformer) This is a bummer because it only gave me one test to install and run our bootloader & softdevice on the device and for some reason it failed - and I couldn't use a debugger to see why.

But when googling I found that apprently the NF51 has a bug in their implementation and if you are tricky it is possible to slowly leak out the contents of flash. So I just wrote a tool (almost entirely based on this great web article): https://github.com/geeksville/nrf51-extractor

It works! The image I pulled out of the new device (which I'm going to carefully leave virgin btw) I could then successfully program onto my development SW102 and it happily now boots the factory load (which I never previously had access to). I can even now turn off the read-protect bit before programming . So I should be able to try our approach and now be able to debug why it didn't work. I bet it will be fairly quick to find the problem now that I can test more than one time per device ;-)

# Update 2
Interesting!  About 25% of the time updating the virginmotized SW102 to our bootloader _works_.  The bootloader successfully comes up and
happily then upgrades to our appload.  75% of the time the uploaded bootloader just seems dead.  Okay, now we can debug!
---
reaching our bootloader!

(gdb) b main
Breakpoint 1 at 0x3cfc4: file ./nRF5_SDK_12.3.0/components/drivers_nrf/hal/nrf_gpio.h, line 514.
(gdb) cont
Continuing.
Note: automatically using hardware breakpoints for read-only addresses.

Breakpoint 1, main () at ./src/main.c:50
50	  nrf_gpio_cfg_input(BUTTON_M__PIN, NRF_GPIO_PIN_PULLUP);
(gdb)

But later we die in ble_stack_init() when it calls sd_softdevice_vector_table_base_set.  Unfortunately the failure is deep inside softdevice
code which I don't have source for.

but doing a raw diff of the entire flash - the programming of our softdevice into the flash seems _very_ slightly corrupted:

```
diff --suppress-common-lines -y <(xxd -e failed-upgrade.bin) <(xxd -e working-scratch-install.bin)  | less
00000300: 4601bd70 04922201 42906848 2009d901  p..F."..Hh.B.. | 00000300: 4601bd70 04122201 42906848 2009d901  p..F."..Hh.B..
00000800: 00000000 00031fe0 00000e40 00000000  ........@..... | 00000800: ffffffff ffffffff ffffffff ffffffff  ..............
``` 

The only diffs are here, the one on the left is the non working version (after upgrading the target using my OTA transitional)
The cols on the right are what they should be.  So I know that the sdk9 bootloader properly installed our bootloader and new softdevice (ALMOST).

I checked the hex softdevice file we built into our image:
arm-none-eabi-objdump -s `find . -name s130_nrf51_2.0.1_softdevice.hex` | less

The contents match what is on the right (as expected)

The values at 0x800 seem to be wrong because that portion of the softdevice.hex file ends right at 0x7ff, so the original rev9 bootloader
chooses not to erase the page at 0x800 (I guess).  

Possibly the failure is due to a single bit error in the 04922201 vs 04122201 (correct) at 0x304.  This is softdevice code and it seems like
the sdk9 bluetooth upgrade made a mistake.

Next step #1: is to manually fix that single bit error and confirm that the board then boots and updates the appload.  This will let me know
that this bit corruption was the cause of the problem.

Next step #2: reinstall the virgin image and re OTA upgrade to our transitional bootloader zip file (and make sure this is one of the 75% failure cases).  Then copy the flash back to my computer and see if there is another single bit error and at a different random address (likely).  
Then dive into sdk9 bootloader source
again to see if there is a way we can add an optional CRC to the files inside the zip - because it seems like something is probably flaky in their
bluetooth download.

**Comment by Gadjet5** [comment]
I have a unit from PSWPower i'm happy to do a test on if it is possible.

**Comment by geeksville** [comment]
Update #3: (see bottom for important bad news and questions) 

did next step #1 try fixing just the messed up word in failed-upgrade.bin, to confirm it works.  Use bless binary editor.  DID WORK!  Therefore the problem is bit level corruption in the upload/flashing process.  

Uploading with NRF Connect worked! our bootloader ran fine.  Trying a few more times after lunch.  Use following commands to revirginmotize the device:

reset halt
nrf51 mass_erase
flash write_image factory-sw102.bin 0
flash write_image factory-sw102-uicr-readable.bin 0x10001000
reset run  

Still fails sometimes: interestingly the bit error is always at the same address and same bit: 0x304.

Test if supply voltage matters (i.e. not using just the little STLINK for power) - no

The verify call to the bootloader over bluetooth succeeds, which means the CRC check of the bootloader+softdevice+MBR showed the bytes were correct.  But looking at the implementation of the old SDK9 bootloader the copy to the final location (not in the application portion of flash
which they use as 'scratch') happens deep in the bowels of the softdevice/MBR.  And it occurs on the NEXT boot, presumbably because they can't write to the MBR while interrupts are running because to write flash first you have to erase pages, and 0xffffffff in your ARM vector table at
address zero would crash things hard.  We don't have source for this 'secret' NordicSemi code, but I presume they copy a special copy routine to RAM and use that routine to erase and write the flash starting at address zero.

It seems to me there is a bug in this (super old) copy routine.  And it is possibly timing/chip specific - one of my SW102s fails only 1 in 10 times, the other fails about 50% of the time.  I bet since people so rarely update softdevices OTA (for good reason), Nordic just missed this
bug.  

In theory we could backport our entire app to the super old SDK9 (so we wouldn't need to update the soft device), but I think that is unwise because there has been a lot of bug fixes since that old release.  I think we pretty much need to stay on SDK12.

So... Bad news: I don't think we can reliably do initial installs over bluetooth with the factory firmware.  Do ya'll think there is any value in me making end-user instructions for how to _try_ to run the update?  I'm on the fence - yes it would save some people from opening their unit
but it would also cause user confusion when it doesn't work and we say "okay - now you need to use an STLINK"

**Comment by geeksville** [comment]
Update #4: Pasting in a good idea @casainho had and then we developed a bit in chat. 
We could check softdevice version in our appload and if it is too old, stop using softdevice entirely.  This would allow installing without opening the device but no bluetooth features.

Copypasta here:

that would be doable - but we'd have to not use any of the stuff that is provided by the softdevice (because the vectors are version specific, we are on S130, the sdk9 was S100 (or 110 - I don't remember off hte top of my head). We currently use softdevice for flash and bluetooth. I'm not sure if flash access is allowed without going through the softdevice (but could check)
I'll try a quick experiment of trying our non softdevice version on the old bootloader tomorrow morning. I already use it when debugging - because softdevice is basically not compatible with the debugger.
oh
I just thought of the problem:
I mentioned this before in the bug but forgot it. The factory bootloader did a really dumb thing - they only support entering the bootloader when requested to by the appload. (Which is super dumb because if you have a DOA appload your device is bricked but whatever). That transition is normally done through a special softdevice call. So maybe we could do something like in our appload at boot check to see if the M and power are held (similar to what we do in our more sane bootloader), if we see that then cobble together a little bit of old SDK9 style softdevice calls by carefully using whatver SVC instruction was used for that rev of the SDK.

The factory appload does this by talking bluetooth and listing a DFU service and if anyone calls that endpoint, they shutdown the appload and let the bootloader handle it - but we'd be running no bluetooth.

well anyways - I'll try the experiment because it sounds easy. But users that go this path would have a bit less margin if we ever pushed out an appload with a serious bug early on.

**Comment by lowPerformer** [comment]
Quite interesting but maybe quite confusing to end-users. But we can explain almost everything on a good wiki page.
So we need in our software

1) A keyless entry mode for the old Bootloader/SD version or users are stuck with just the version they flashed the first time.

2) to be sure flash writes are doable without SD interaction. But afaik all flash writes are managed by the Softdevice (at least with the SDK we use).

**Comment by geeksville** [comment]
@lowPerformer yep.  I haven't looked at how flash writing works, but if it needs softdevice this idea might be DOA.  I just tried a quick experiment to see how easily we could run our app without using the softdevice (and I've added this as a standard feature in master).  if useSoftDevice is set to false we don't do bluetooth and we don't do flash and the app basically works.

I then changed the start addr of the appload in gcc_nrf51.ld to match what SDK9 would expect (0x18000) and flashed on a factory image (with their bootloader/softdevice/app).  I was then able to install and run our app on the factory device without opening the case.

Commit here: https://github.com/geeksville/Color_LCD/commit/e6a7e5437b469e927a562c502a52953a6967f827

Alas, tomorrow night I'm leaving California to Japan for a few weeks.  So if you or anyone else wants to keep going on this I can comment here or in gitter, otherwise I'll resume this when I return.

**Comment by casainho** [comment]
Are you sure there isn't a way to write to flash/ save configurations without using the softdevice?

And even if it is a must, why can't we use it, I mean, that very old version??

**Comment by eyebyesickle** [comment]
So, to be clear, with the partial success factor, are we talking sometimes it works, but if not, the display is bricked without soldering/st link? Or what?

There is also at least the option to have US and Chinese/INTL stock to distribute to everyone preloaded from the factory... they are willing to do that, and I can have them priced the same as PSW power, which is where most seem to buy...

Mind shooting me a file to play around with regardless? I had cracking em open, and wouldn't mind at least getting one loaded.

Also, to clarify (since I am an idiot) - once you do have success - then future updates are a breeze, correct?

I think no matter what, we can get something workable... even if we just flood the market with preloaded versions matching whatever the cheapest price is for people... at least it would be convenient in the meantime of searching for a better way...

Just my 2 cents - which of course comes from my limited understanding! hehe

**Comment by geeksville** [comment]
> Are you sure there isn't a way to write to flash/ save configurations without using the softdevice?
Nope.  Haven't read the datasheet on the chip to see if flash writes is allowed from appspace without softdevice.  And if allowed, what the algorithm is (or see if someone else has already written it).  That investigation would take a few hours.

> And even if it is a must, why can't we use it, I mean, that very old version??
Without the ability to write flash we have no way to save settings (which is critical, considering the display provisions the motor based on those settings)

**Comment by geeksville** [comment]
> So, to be clear, with the partial success factor, are we talking sometimes it works, but if not, the display is bricked without soldering/st link? Or what?

Yes, if it fails the critical 'softdevice' firmware from nordic semi is corrupted.  If that happens, the user would then need to open the device and install using stlink (which we have setup to always rewrite everything with valid stuff).

> 
> There is also at least the option to have US and Chinese/INTL stock to distribute to everyone preloaded from the factory... they are willing to do that, and I can have them priced the same as PSW power, which is where most seem to buy...

That would be AMAZING.  Yes, if they installed our bootloader/softdevice/app image EVERYTHING is easy.  And I'm sure everyone that uses this project would buy your preconfigured devices.   None of this trickery in this thread would be needed.

> 
> Mind shooting me a file to play around with regardless? I had cracking em open, and wouldn't mind at least getting one loaded.

I don't quite understand.  Do you want me to send you this hacked up version of our app that can be installed on their bootloader over the air?  Could do - but just so you know, it isn't really usable because settings saving had to be disabled for this experiment.  Also, it doesn't have the magic glue to ever let you restart their bluetooth software update process.  So I don't think you want to put this on a device.

> Also, to clarify (since I am an idiot) - once you do have success - then future updates are a breeze, correct?

Yes, if using this hack to use our app on their old bootloader/softdevice, future software updates of our app (or even going back to their app) would work.  But we'd need to do a bit of work described above.  My little quick test was just a proof of concept.  It is missing the flash writing code we depend on, and a place where we jump back into their old bootloader if the user wants to update software.

> I think no matter what, we can get something workable... even if we just flood the market with preloaded versions matching whatever the cheapest price is for people... at least it would be convenient in the meantime of searching for a better way...

IMO - preloaded with our app/bootloader if you have a vendor who would do it is the best of all worlds.  If you do want to do this, give us a little advance notice because I think we should make a couple of small tweaks to our bootloader before we give them an official image to burn in.  The only change I think we should do is to set a PRODUCT and VENDOR code for our 'product' - currently both the factory bootloader and our bootloader leave it at the default of 0xff.  If we set it to something we can then make a nicer user experience for future software updates (so users can't accidentally put bogus loads onto a device).


> Just my 2 cents - which of course comes from my limited understanding! hehe

**Comment by lowPerformer** [comment]
@geeksville @casainho 
SDK documentation is very clear about flash write and SD:
https://infocenter.nordicsemi.com/topic/com.nordic.infocenter.sdk5.v12.3.0/lib_fstorage.html

We use Flash Data Storage, which interacts with SD because of timing concerns. I also can remember that indeed writes didn't work first because of wrong SD init.

Even worse, SDK 9 doesent have those Flash Storage / Flash Data Storage libraries at all. It uses very basic direct writes instead:
https://infocenter.nordicsemi.com/topic/com.nordic.infocenter.sdk51.v9.0.0/flashwrite_example.html

which maybe is good?
SDK12.3 w Bluetooth -> use fancy Flash Data Storage
SDK9.0 w/o Bluetooth -> use basic Flash Writes

**Comment by casainho** [comment]
Well, on 850C I am using that basic flash writes for configurations and works ok as expected, so, maybe the code can be reused on SW102.

**Comment by geeksville** [comment]
@casainho alas no.  The problem is not the fancy flash filesystem, it is even the raw flash read/write is not done by a lib on NRF51, but by the softdevice.  We would need to either find a lib or read the datasheet for the CPU and implement the flash erase/program algorithm.

**Comment by lowPerformer** [comment]
@geeksville @casainho 
With SDK9 it is a raw flash write just by configuring the NVMC and writing to the address.
From the SDK9 flash_write example:
```
/** @brief Function for filling a page in flash with a value.
 *
 * @param[in] address Address of the first word in the page to be filled.
 * @param[in] value Value to be written to flash.
 */
static void flash_word_write(uint32_t * address, uint32_t value)
{
    // Turn on flash write enable and wait until the NVMC is ready:
    NRF_NVMC->CONFIG = (NVMC_CONFIG_WEN_Wen << NVMC_CONFIG_WEN_Pos);

    while (NRF_NVMC->READY == NVMC_READY_READY_Busy)
    {
        // Do nothing.
    }

    *address = value;

    while (NRF_NVMC->READY == NVMC_READY_READY_Busy)
    {
        // Do nothing.
    }

    // Turn off flash write enable and wait until the NVMC is ready:
    NRF_NVMC->CONFIG = (NVMC_CONFIG_WEN_Ren << NVMC_CONFIG_WEN_Pos);

    while (NRF_NVMC->READY == NVMC_READY_READY_Busy)
    {
        // Do nothing.
    }
}
```

So even if this all factory OTA is really a mess we still can do two versions:
1) w/o Bluetooth but OTA upgradeable, no need to open.
2) with all fancy stuff but with ~50 % chance to have to open it and flash by ST-Link

**Comment by lowPerformer** [comment]
But to maintain two quite different versions of SW102 software is also quite a pain!

**Comment by casainho** [comment]
So I am happy to have a possible solution without the need to open it.

Maybe we can use the same flash write for all versions ??

**Comment by geeksville** [comment]
Thanks for looking!  Sounds good then and not super hard.  If someone wants
to do it before I get back then awesome!  Otherwise I'll work on it when I
get home on Nov 5th.

I don't even think we'd have to keep two separate versions.   Just a build
flag set in the makefile saying FOROLDSDK.  Based on that flag:

* we'd change the start addr for the appload
* Set useSoftDevice=false (I just added this flag for my quick test)
* Call handrolled version of flash read/write
* Very early call possiblyEnterBootloader().  That function would wait for
the realease of the power button, if the user holds the pretty button for
>10 seconds while booting we branch into the sdk9 bootloader and say user
might want a DFU.
* Put the previous call as early as possible to decrease the odds we ever
put a fatal but before that point. 😃

(Sent from a phone - please ignore typos)

**Comment by casainho** [comment]
I want to focus myself on 850C. I need a version with the same functionalities as the old firmware, so finally I can switch to it. Then I will be able to work in both.

**Comment by eyebyesickle** [comment]
You know, you may be able to reach out to Eggrider and ask them how they fixed the issue... Just a thought. I don't have the best relations with them so this would be a task for someone else, lol

**Comment by geeksville** [comment]
@eyebyesickle did they fix the issue?  do you have an email address there?  I can send them a note.  

Btw - I got back into town a few days ago.  Unless @casainho objects I'm going to go do the hack we described above this week.

**Comment by casainho** [comment]
I am a bit lost but I know that would be great if users could run our firmware without need to open the SW102 (even if it will not be possible to use Bluetooth communications with a mobile app, that we don´t have now anyway).

850C is currently the recommended display but then, SW102 would be it.

**Comment by geeksville** [comment]
Good progress with our hack 'not a fully featured app but pretty good' version.  I can now install our app on devices without opening the case.  I've also modified our app so that with the appropriate button press the user has a way to force the bootloader to wait for updates.

The only remaining item is to change the flash writing, which I'll do tomorrow.

**Comment by geeksville** [comment]
Update: my WIP branch can happily be installed on devices now without opening them.  And almost all features of the regular build (no bluetooth) are supported and work identically.

Alas - I can't release yet because the second software update installed via bluetooth doesn't work (I need to hack some stuff to work with the old shitty bootloader).  

Unfortunately, I'm out of town next week, but I'll bring my SW102 with me to continue development.  I'll probably be able to make a release sometime next week...

**Comment by geeksville** [comment]
Alas - I underestimated the various tasks involved in getting my mom moved in with my brother for her post stroke life.  Realistically, I won't be doing any programming until I get back to California on the 3rd.

**Comment by casainho** [comment]
Family first!!

**Comment by cookejames** [comment]
@geeksville I was wondering if you had a branch that I could test this with? I have an unopened SW102 that I was wanting to try this on and was interested in looking at the code.

**Comment by geeksville** [comment]
@cookejames  Sure.  Alas, I don't think I'll be getting back to this for a while.  If you wanna keep working on it that's great!

https://github.com/geeksville/Color_LCD/tree/real-nosolder 

Current status:
This branch if built with "make release" (notice in the Makefile this really just means "make NOSOLDER=true clean generate_nosolder") will generate sw102-nosolder-version.zip.  
This zip file will be accepted by the factory bootloader and be installed by on the device and work correctly with essentially all of our features work.

The problem is: When you want to update _this_ appload.  You'll notice I have code in main.c to branch back into the factory bootloader when the user requests a new software update.  The branch to the bootloader happens correctly (and note - we don't have source to this bootloader but I assume with high confidence that it is essentially the example code bootloader distributed with SDK rev 9).  But once in the bootloader it seems to drop the bluetooth connection.  I think the error is we need to have our appload call bluetooth init (because dumbly, the sdk9 bootloader seems to expect that).

**Comment by cookejames** [comment]
Ok thanks for letting me know. I was hoping to not have to open up my SW102 but it seems like I'd only have one shot at fixing the bootloading code and then I would have to open it anyway which is annoying as otherwise I would have given it a go.

I guess I'll open up mine and if I have the time I'll have a go at fixing this problem.

**Comment by obcd007** [comment]
Besides jumping from firmware 2 bootloader, is there another way to keep the unit into the bootloader so the firmware can be updated using bluetooth? Sorry if the answer is obvious. I haven't studied the bootloader or firmware code yet.

**Comment by yodapotatofly** [comment]
Hello there ! I'm very interested by this future improvment !

As @geeksville is no longer assigned on this, I assume that no one is working on it anymore ?

I don't have technica knowledge on this, so I'm kinda lost with what you guys are talking about. Could you please indicate how close are we (or not) to getting it to work ?

Also, have I understood correctly that once it will be successfully implemented, only devices sold by @eyebyesickle (Eco Cycles) or EggRider would work, because they allow flashing firmware through bluetooth ?

Thanks a lot !
